// TLP.Example02/index.js
"use strict";
var Control_Category = require("../Control.Category/index.js");
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Data_Show = require("../Data.Show/index.js");
var Prelude = require("../Prelude/index.js");
var A = (function () {
    function A(value0) {
        this.value0 = value0;
    };
    A.create = function (value0) {
        return new A(value0);
    };
    return A;
})();
var B = (function () {
    function B(value0) {
        this.value0 = value0;
    };
    B.create = function (value0) {
        return new B(value0);
    };
    return B;
})();
var main$prime = function (sometype) {
    if (sometype instanceof A) {
        return Control_Category.identity(Control_Category.categoryFn)(sometype.value0);
    };
    if (sometype instanceof B) {
        return Control_Category.identity(Control_Category.categoryFn)(sometype.value0);
    };
    throw new Error("Failed pattern match at TLP.Example02 line 12, column 18 - line 14, column 28: " + [ sometype.constructor.name ]);
};
var main = function (v) {
    if (v instanceof A) {
        return Control_Category.identity(Control_Category.categoryFn)(v.value0);
    };
    if (v instanceof B) {
        return Control_Category.identity(Control_Category.categoryFn)(v.value0);
    };
    throw new Error("Failed pattern match at TLP.Example02 line 7, column 1 - line 7, column 27: " + [ v.constructor.name ]);
};
var someFunction = main(new A("5"));
var foo = function (dictShow) {
    return function (x) {
        return function (s) {
            return Data_Show.show(dictShow)(x) + s;
        };
    };
};
var bar = function (dictShow) {
    return function (y) {
        return Data_Semigroup.append(Data_Semigroup.semigroupFn(Data_Semigroup.semigroupFn(Data_Semigroup.semigroupString)))(y)(foo(dictShow));
    };
};
module.exports = {
    A: A,
    B: B,
    main: main,
    "main'": main$prime,
    someFunction: someFunction,
    foo: foo,
    bar: bar
};
